apiserver/facades/agent/uniter/networkinfo.go

```diff
@@ -165,6 +138,60 @@ func (n *NetworkInfoBase) maybeGetUnitAddress(rel *state.Relation) (corenetwork.
 	return nil, nil
 }
 
+// getEgressForRelation returns any explicitly defined egress subnets
+// for the relation, falling back to configured model egress.
+// If there are none, it attempts to resolve a subnet from the input
+// ingress addresses.
+func (n *NetworkInfoBase) getEgressForRelation(
+	rel *state.Relation, ingress corenetwork.SpaceAddresses,
+) ([]string, error) {
+	egressSubnets, err := state.NewRelationEgressNetworks(n.st).Networks(rel.Tag().Id())
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	if egressSubnets != nil {
+		egress := egressSubnets.CIDRS()
+		if len(egress) > 0 {
+			return egress, nil
+		}
+	}
+
+	if len(n.defaultEgress) > 0 {
+		return n.defaultEgress, nil
+	}
+
+	egress, err := n.getEgressFromIngress(ingress.Values())
+	return egress, errors.Trace(err)
+}
+
+// getEgressFromIngress returns a subnet corresponding to the first address
+// (if available) in the input ingress address list.
+// There can be situations (observed for CAAS) where the preferred ingress
+// address is a FQDN for a load-balancer, which is intended to point at a
+// service that is not yet up.
+// We employ the retry strategy here to give time for the FQDN to be resolvable
+// to an IP address.
+func (n *NetworkInfoBase) getEgressFromIngress(ingress []string) ([]string, error) {
+	if len(ingress) == 0 {
+		return nil, nil
+	}
+
+	var egress []string
+	retryArg := n.retryFactory()
```

Comment:
```golang
var defaultRetryFactory = func() retry.CallArgs {
	return retry.CallArgs{
		Clock:       clock.WallClock,
		Delay:       3 * time.Second,
                 // 30s is unfortunately not enough for AWS LB to be fully provisioned.
                  // It could take a few mins.
                  // We probably should not retry, ignore `DNSError` and just return empty egress to let 
                  // client retry later.
		MaxDuration: 30 * time.Second, 
	}
}
```